memgpt/openai_tools.py

```diff
@@ -75,6 +81,34 @@ def completions_with_backoff(**kwargs):
         return openai.ChatCompletion.create(**kwargs)
 
 
+@retry_with_exponential_backoff
+def chat_completion_with_backoff(agent_config: AgentConfig, **kwargs):
+    from memgpt.utils import printd
+
+    printd(f"Using model {agent_config.model_endpoint_type}, endpoint: {agent_config.model_endpoint}")
+    if agent_config.model_endpoint_type == "openai":
+        # openai
+        openai.api_base = agent_config.model_endpoint
+        return openai.ChatCompletion.create(**kwargs)
+    elif agent_config.model_endpoint_type == "azure":
+        # configure openai
+        config = MemGPTConfig.load()  # load credentials (currently not stored in agent config)
+        openai.api_type = "azure"
+        openai.api_key = config.azure_key
+        openai.api_base = config.azure_endpoint
+        openai.api_version = config.azure_version
+        if config.azure_deployment is not None:
+            kwargs["deployment_id"] = config.azure_deployment
+        else:
+            kwargs["engine"] = MODEL_TO_AZURE_ENGINE[config.model]
+            del kwargs["model"]
+        return openai.ChatCompletion.create(**kwargs)
+    else:  # local model

```

Comment:
Need to pull `BACKEND_TYPE` from `endpoint_type`:

```
  File "/Users/loaner/dev/MemGPT-swooders/memgpt/openai_tools.py", line 108, in chat_completion_with_backoff
    return get_chat_completion(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/loaner/dev/MemGPT-swooders/memgpt/local_llm/chat_completion_proxy.py", line 39, in get_chat_completion
    raise ValueError(f"The BACKEND_TYPE environment variable is not defined. Please set it in your environment.")
ValueError: The BACKEND_TYPE environment variable is not defined. Please set it in your environment.
```

Reply:
Sorry I mean pull then pass to `chatcompletion_proxy` (I can add this)

Reply:
Actually, @sarahwooders do you think it makes more sense to pass these through to the `get_chat_completion` call as kwargs, or for `chat_completion_proxy.py` to pull them from the config? The former right.

<img width="610" alt="image" src="https://github.com/cpacker/MemGPT/assets/5475622/10c989ba-fb82-4681-b815-30776fbbf11d">

Reply:
patched here: https://github.com/cpacker/MemGPT/pull/422/commits/50f79e5953178abafc453f9e41223a6e31424485

Reply:
I think either way is fine - I'd kind of prefer to get rid of the the `**kwargs` and just pass in the fields needed to each chat completion function from `chat_completion_with_backoff(config, messages, functions, function_call)`